### PR TITLE
[dev-launcher] add dev sessions and update deep link modal

### PR DESCRIPTION
**Pull request by ajsmth**
# Why

This PR updates the "local packager" names to be a more accurate "dev session" and adds support for querying dev sessions (for authenticated users), adding the recently opened apps section to the Home Screen,  and improved handling of deep links via modal

Screen Shot 2021-12-15 at 5 07 57 PM

Screen Shot 2021-12-15 at 12 07 30 PM

I've also updated the url input validation to be throttled, and to fire only if there is already an error w/ the url and on input blur - this is a better UX IMO

@esamelson this will likely clash with your current PR that adds the device ID to the dev session query - I can fix this PR up after yours is merged to make sure its all good!

I've also included a button to bring up a prompt that instructs people how to start a dev server - this is similar to the empty state of the packager list, but is available in case someone is confused about how to get a server to show up in the list: 

Screen Shot 2021-12-15 at 5 23 47 PM

Screen Shot 2021-12-15 at 5 21 34 PM

# How

### Dev Sessions:

Dev sessions are querying on app launch and via refetch - they make use of the `DevMenu.queryDevSessions` method and are mapped to JS and displayed

For the case of simulator builds, the app will also attempt to fetch `localhost` packagers 
The deep link modal has some improved styles, and includes all available dev servers to be directly opened

# Test Plan

I've added unit tests to ensure `queryDevSessions` is called and the results are rendered, and `getRecentlyOpenedApps` is called and the results are rendered 

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-dev-launcher/bundle/hooks/useRecentlyOpenedApps.tsx [hunk 1]
```diff
@@ -44,6 +44,7 @@ export function useRecentlyOpenedApps() {
           };
         });
 
+
```
Comment (lukmccall) [inline packages/expo-dev-launcher/bundle/hooks/useRecentlyOpenedApps.tsx:47]:
```suggestion
```
🧹

File: packages/expo-dev-launcher/bundle/native-modules/DevLauncherInternal.ts [hunk 2]
```diff
@@ -43,3 +43,5 @@ export async function copyToClipboardAsync(content: string): Promise<null> {
 }
 
 export const clientUrlScheme = DevLauncher.clientUrlScheme;
+
+export const isDevice = !!DevLauncher.isDevice;
```
Comment (lukmccall) [inline packages/expo-dev-launcher/bundle/native-modules/DevLauncherInternal.ts:47]:
```suggestion
export const isDevice = !!DevLauncher.isDevice;

```

File: packages/expo-dev-launcher/bundle/components/redesign/DeepLinkModal.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  Divider,
+} from 'expo-dev-client-components';
+import * as React from 'react';
+import { Alert, ScrollView } from 'react-native';
+
+import { useDevSessions } from '../../hooks/useDevSessions';
+import { useRecentlyOpenedApps } from '../../hooks/useRecentlyOpenedApps';
+import { loadApp } from '../../native-modules/DevLauncherInternal';
+
+export function DeepLinkModal({ onClosePress, pendingDeepLink }) {
```
Comment (lukmccall) [inline packages/expo-dev-launcher/bundle/components/redesign/DeepLinkModal.tsx:20]: Could you split this component into smaller ones? Currently, it's hard to understand what going on there.

File: packages/expo-dev-launcher/bundle/components/redesign/Modal.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  React.useEffect(() => {
+    if (hasModal) {
+      Animated.spring(animatedValue.current, {
+        toValue: 1,
+        useNativeDriver: false,
+      }).start();
+    }
+
+    if (!hasModal) {
```
Comment (lukmccall) [inline packages/expo-dev-launcher/bundle/components/redesign/Modal.tsx:43]:
```suggestion
    } else {
```

**Review by lukmccall — CHANGES_REQUESTED** [review]
Just a couple of suggestions ;)

File: packages/expo-dev-launcher/bundle/screens/HomeScreen.tsx [hunk 5]
```diff
@@ -246,3 +270,45 @@ function RecentlyOpenedApps({ onAppPress }) {
...
+
+        <View bg="secondary" border="default" rounded="medium" padding="medium">
+          <Text type="mono">expo start</Text>
+        </View>
+
+        <Spacer.Vertical size="large" />
+        <Text>Then, select the local server when it appears here.</Text>
+        <Spacer.Vertical size="small" />
+        <Text>
+          Alternatively, open the Camera app and scan the QR code that appears in your terminal
```
Comment (esamelson) [inline packages/expo-dev-launcher/bundle/screens/HomeScreen.tsx:309]:
```suggestion
          Alternatively, open the Camera app and scan the QR code that appears in your terminal.
```
tiny nit, just to match the rest of the text :)

**Review by esamelson — APPROVED** [review]
looks great! 🎉

**Review by lukmccall — APPROVED** [review]

**Pull request merged**